# Latent Dirchlet Allocation (LDA)

In [1]:
import pandas as pd

In [2]:
npr = pd.read_csv('npr.csv')

### NPR Radio Transcripts conversation.

In [3]:
npr.head()

Article
0  In the Washington of 2016, even when the polic...
1    Donald Trump has used Twitter  —   his prefe...
2    Donald Trump is unabashedly praising Russian...
3  Updated at 2:50 p. m. ET, Russian President Vl...
4  From photography, illustration and video, to d...

In [4]:
npr['Article'][0]

'In the Washington of 2016, even when the policy can be bipartisan, the politics cannot. And in that sense, this year shows little sign of ending on Dec. 31. When President Obama moved to sanction Russia over its alleged interference in the U. S. election just concluded, some Republicans who had long called for similar or more severe measures could scarcely bring themselves to approve. House Speaker Paul Ryan called the Obama measures ”appropriate” but also ”overdue” and ”a prime example of this administration’s ineffective foreign policy that has left America weaker in the eyes of the world.” Other GOP leaders sounded much the same theme. ”[We have] been urging President Obama for years to take strong action to deter Russia’s worldwide aggression, including its   operations,” wrote Rep. Devin Nunes,  . chairman of the House Intelligence Committee. ”Now with just a few weeks left in office, the president has suddenly decided that some stronger measures are indeed warranted.” Appearing 

### Checking Length of Dataset

In [5]:
len(npr)

11992

### Preprocesssing before LDA Process

In [6]:
from sklearn.feature_extraction.text import CountVectorizer

**`max_df`**` : float in range [0.0, 1.0] or int, default=1.0`<br>
When building the vocabulary ignore terms that have a document frequency strictly higher than the given threshold (corpus-specific stop words). If float, the parameter represents a proportion of documents, integer absolute counts. This parameter is ignored if vocabulary is not None.

**`min_df`**` : float in range [0.0, 1.0] or int, default=1`<br>
When building the vocabulary ignore terms that have a document frequency strictly lower than the given threshold. This value is also called cut-off in the literature. If float, the parameter represents a proportion of documents, integer absolute counts. This parameter is ignored if vocabulary is not None.

### Doing count vectorizer and removing stop words from Articles.

In [7]:
cv = CountVectorizer(max_df=0.95, min_df=2, stop_words='english')

** Discard word that show up 95 percent of the documents.

In [8]:
doc = cv.fit_transform(npr['Article'])

In [9]:
doc

<11992x54777 sparse matrix of type '<class 'numpy.int64'>'
	with 3033388 stored elements in Compressed Sparse Row format>

In [10]:
from sklearn.decomposition import LatentDirichletAllocation

In [12]:
LDA = LatentDirichletAllocation(n_components=7,random_state=42)

In [13]:
LDA.fit(doc)

LatentDirichletAllocation(n_components=7, random_state=42)

### Grab Vocabulary of words

** Count vectorizer element is dimension of per document * words.

In [14]:
len(cv.get_feature_names())

54777

** Bags of word at index 5000

In [31]:
cv.get_feature_names()[5000]

'bask'

In [21]:
#randomly grab indexes 

import random
rand_id = random.randint(0,54777)
cv.get_feature_names()[rand_id]

'bighorn'

In [22]:
len(LDA.components_)

7

** LDA component as declared in initialization.

In [23]:
LDA.components_

array([[8.64332806e+00, 2.38014333e+03, 1.42900522e-01, ...,
        1.43006821e-01, 1.42902042e-01, 1.42861626e-01],
       [2.76191749e+01, 5.36394437e+02, 1.42857148e-01, ...,
        1.42861973e-01, 1.42857147e-01, 1.42906875e-01],
       [7.22783888e+00, 8.24033986e+02, 1.42857148e-01, ...,
        6.14236247e+00, 2.14061364e+00, 1.42923753e-01],
       ...,
       [3.11488651e+00, 3.50409655e+02, 1.42857147e-01, ...,
        1.42859912e-01, 1.42857146e-01, 1.42866614e-01],
       [4.61486388e+01, 5.14408600e+01, 3.14281373e+00, ...,
        1.43107628e-01, 1.43902481e-01, 2.14271779e+00],
       [4.93991422e-01, 4.18841042e+02, 1.42857151e-01, ...,
        1.42857146e-01, 1.43760101e-01, 1.42866201e-01]])

### Single Topic.

In [24]:
single_topic = LDA.components_[0]

In [25]:
single_topic.argsort()

array([ 2475, 18302, 35285, ..., 22673, 42561, 42993])

### Grab highest probability word per topic

In [27]:
# Grab Top 10 Greatest Value
# Last 10 Value from LDA components index 0
single_topic.argsort()[-10:]

array([33390, 36310, 21228, 10425, 31464,  8149, 36283, 22673, 42561,
       42993])

In [35]:
top_ten_words = single_topic.argsort()[-10:]
for i in top_ten_words:
    print(cv.get_feature_names()[i])

new
percent
government
company
million
care
people
health
said
says


** Collection of Top 10 word in single topic (Topic 0)

### Printing Per-Topic Top Words.

In [38]:
for idx,topic in enumerate(LDA.components_):
    print(f"The Top 10 Words for Topic {idx}")
    print([cv.get_feature_names()[index] for index in topic.argsort()[-10:]])
    print('\n')

The Top 10 Words for Topic 0
['new', 'percent', 'government', 'company', 'million', 'care', 'people', 'health', 'said', 'says']


The Top 10 Words for Topic 1
['npr', 'reports', 'says', 'news', 'people', 'told', 'police', 'president', 'trump', 'said']


The Top 10 Words for Topic 2
['time', 'water', 'city', 'new', 'years', 'food', 'just', 'people', 'like', 'says']


The Top 10 Words for Topic 3
['disease', 'patients', 'just', 'children', 'study', 'like', 'women', 'health', 'people', 'says']


The Top 10 Words for Topic 4
['obama', 'court', 'republican', 'campaign', 'people', 'state', 'president', 'clinton', 'said', 'trump']


The Top 10 Words for Topic 5
['new', 'way', 'music', 'really', 'time', 'know', 'think', 'people', 'just', 'like']


The Top 10 Words for Topic 6
['people', 'time', 'schools', 'just', 'education', 'new', 'like', 'students', 'school', 'says']




### Assign Topic Number into Dataframe

1. Declaring Count Vectorizer Method.
2. Fit and Transform Article of each column in dataframe
3. Declaring LDA
4. Fit Count Vectorizer into LDA method

In [41]:
doc

<11992x54777 sparse matrix of type '<class 'numpy.int64'>'
	with 3033388 stored elements in Compressed Sparse Row format>

** no need to do count vectorizer and stop words again.

In [39]:
topic_num = LDA.fit_transform(doc)

In [42]:
topic_num.shape

(11992, 7)

** Contain 11992 Article with each 7 Topics has vary probability, thus the highest probability index will be its topic number (0-7).

In [46]:
topic_num[0].round(2)

array([0.02, 0.68, 0.  , 0.  , 0.3 , 0.  , 0.  ])

** shape of 'topic_num' topic index 1 has highest prob.

In [40]:
topic_num[0].argmax()

1

In [43]:
npr['Topic'] = topic_num.argmax(axis=1)

In [48]:
npr.head(10)

Article  Topic
0  In the Washington of 2016, even when the polic...      1
1    Donald Trump has used Twitter  —   his prefe...      1
2    Donald Trump is unabashedly praising Russian...      1
3  Updated at 2:50 p. m. ET, Russian President Vl...      1
4  From photography, illustration and video, to d...      2
5  I did not want to join yoga class. I hated tho...      3
6  With a   who has publicly supported the debunk...      3
7  I was standing by the airport exit, debating w...      2
8  If movies were trying to be more realistic, pe...      3
9  Eighteen years ago, on New Year’s Eve, David F...      2